src/indexer/index_writer.rs

```diff
@@ -643,12 +647,45 @@ impl IndexWriter {
     pub fn add_document(&mut self, document: Document) -> u64 {
         let opstamp = self.stamper.stamp();
         let add_operation = AddOperation { opstamp, document };
-        let send_result = self.document_sender.send(add_operation);
+        let send_result = self.document_sender.send(vec![add_operation]);
         if let Err(e) = send_result {
             panic!("Failed to index document. Sending to indexing channel failed. This probably means all of the indexing threads have panicked. {:?}", e);
         }
         opstamp
     }
+
+    /// Runs a group of document operations.
+    pub fn run(&mut self, user_operations: &[UserOperation]) -> Option<u64> {
+        let mut stamps = self.stamper.stamps(user_operations.len() as u64);
+        let mut adds: Vec<AddOperation> = Vec::new();
+        let mut last_opstamp: Option<u64> = None;
+
+        for user_op in user_operations {
+            let opstamp = stamps.next().unwrap();
+            last_opstamp = Some(opstamp);
+            match user_op {
+                UserOperation::Delete(term) => {
+                    let delete_operation = DeleteOperation {
+                        opstamp: opstamp,
+                        term: term.clone(),
+                    };
+                    self.delete_queue.push(delete_operation);
+                }
+                UserOperation::Add(doc) => {
+                    let add_operation = AddOperation {
+                        opstamp: opstamp,
+                        document: doc.clone(),
```

Comment:
I don't think we need to clone here do we?

Reply:
I think using `Vec<UserOperation>` is a better alternative. It is unlikely that the `UserOperation`s that are passed to `IndexWriter.run` will be needed by the caller after. I'll make the changes.